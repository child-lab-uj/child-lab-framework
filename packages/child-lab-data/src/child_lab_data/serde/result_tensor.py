import torch
from beartype import beartype
from jaxtyping import Float
from plum import dispatch, overload
from vpc import gaze, pose


@beartype
@overload
def serialize(item: pose.Result3d) -> Float[torch.Tensor, 'n_detections 74']:  # noqa: F811
    flat_boxes = item.boxes  # n_detections x 6
    flat_keypoints = item.keypoints.flatten(1, -1)  # n_detections x 17 * 4
    return torch.cat((flat_boxes, flat_keypoints), dim=1)


@beartype
@overload
def deserialize(  # noqa: F811
    ty: type[pose.Result3d],
    item: Float[torch.Tensor, 'n_detections 74'],
) -> pose.Result3d:
    boxes = item[:, :6]
    keypoints = item[:, 6:].unflatten(-1, (17, 4))
    return pose.Result3d(boxes, keypoints, item.device)


@beartype
@overload
def serialize(item: pose.Result) -> Float[torch.Tensor, 'n_detections 56']:  # noqa: F811
    flat_boxes = item.boxes  # n_detections x 5
    flat_keypoints = item.keypoints.flatten(1, -1)  # n_detections x 17 * 3
    return torch.cat((flat_boxes, flat_keypoints), dim=1)


@beartype
@overload
def deserialize(  # noqa: F811
    ty: type[pose.Result],
    item: Float[torch.Tensor, 'n_detections 56'],
) -> pose.Result:
    boxes = item[:, :6, :]
    keypoints = item[:, 6:, :].unflatten(-1, (17, 3))
    return pose.Result(boxes, keypoints, item.device)


@beartype
@overload
def serialize(item: gaze.Result3d) -> Float[torch.Tensor, 'n_detections 4 3']:  # noqa: F811
    return torch.cat((item.eyes, item.directions), dim=1)


@beartype
@overload
def deserialize(  # noqa: F811
    ty: type[gaze.Result3d],
    item: Float[torch.Tensor, 'n_detections 4 3'],
) -> gaze.Result3d:
    eyes = item[:, :2, :]
    directions = item[:, 2:, :]
    return gaze.Result3d(eyes, directions, item.device)


@dispatch
def serialize(item: object) -> torch.Tensor: ...  # noqa: F811


@dispatch
def deserialize(ty: type, item: object) -> object: ...  # noqa: F811
