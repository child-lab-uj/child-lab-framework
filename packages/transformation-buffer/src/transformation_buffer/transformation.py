from dataclasses import dataclass
from typing import Literal, Self

import serde
import torch
from jaxtyping import Float

from transformation_buffer.serialization.tensor import (
    deserialize_tensor,
    serialize_tensor,
)

type Axis = Literal['x', 'y', 'z']


def rotation_around(
    axis: Axis,
    angle: float | torch.Tensor,
) -> Float[torch.Tensor, '3 3']:
    a = torch.tensor(angle)
    sin = torch.sin(a)
    cos = torch.cos(a)

    match axis:
        case 'x':
            return torch.tensor(
                [
                    [1.0, 0.0, 0.0],
                    [0.0, cos, -sin],
                    [0.0, sin, cos],
                ]
            )

        case 'y':
            return torch.tensor(
                [
                    [cos, 0.0, sin],
                    [0.0, 1.0, 0.0],
                    [-sin, 0.0, cos],
                ]
            )

        case 'z':
            return torch.tensor(
                [
                    [cos, -sin, 0.0],
                    [sin, cos, 0.0],
                    [0.0, 0.0, 1.0],
                ]
            )


def translation_along(axis: Axis, distance: float) -> Float[torch.Tensor, '3']:
    t = torch.zeros(3)

    match axis:
        case 'x':
            t[0] = distance
        case 'y':
            t[1] = distance
        case 'z':
            t[2] = distance

    return t


@serde.serde
@dataclass(eq=False)
class Transformation:
    rotation_and_translation: Float[torch.Tensor, '4 4'] = serde.field(
        serializer=serialize_tensor,
        deserializer=deserialize_tensor,
    )

    def __init__(
        self,
        rotation_and_translation: Float[torch.Tensor, '4 4'],
    ) -> None:
        self.rotation_and_translation = rotation_and_translation.clone()

    def __eq__(self, other: object, /) -> bool:
        return isinstance(other, Transformation) and bool(
            self.rotation_and_translation.eq(other.rotation_and_translation).all()
        )

    def __matmul__(self, other: 'Transformation') -> 'Transformation':
        return Transformation(
            other.rotation_and_translation @ self.rotation_and_translation
        )

    def inverse(self) -> 'Transformation':
        return Transformation(self.rotation_and_translation.inverse())

    def rotation(self) -> Float[torch.Tensor, '3 3']:
        return self.rotation_and_translation[:3, :3].clone()

    def rotation_quaternion(self) -> Float[torch.Tensor, '4']:
        # https://en.wikipedia.org/wiki/Rotation_matrix#Quaternion

        rotation = self.rotation_and_translation

        t = rotation[0, 0] + rotation[1, 1] + rotation[2, 2]

        if t > 0.0:
            r = torch.sqrt(1.0 + t)
            s = 1 / (2 * r)
            w = r / 2
            x = (rotation[2, 1] - rotation[1, 2]) * s
            y = (rotation[0, 2] - rotation[2, 0]) * s
            z = (rotation[1, 0] - rotation[0, 1]) * s
        else:
            r = torch.sqrt(1.0 + rotation[0, 0] - rotation[1, 1] - rotation[2, 2])
            s = 1 / (2 * r)
            w = (rotation[2, 1] - rotation[1, 2]) * s
            x = r / 2
            y = (rotation[0, 1] + rotation[1, 0]) * s
            z = (rotation[0, 2] + rotation[2, 0]) * s

        result = torch.stack((w, x, y, z))

        return result

    def translation(self) -> Float[torch.Tensor, '3']:
        return self.rotation_and_translation[:3, 3].clone()

    @staticmethod
    def approx_eq(
        t1: 'Transformation',
        t2: 'Transformation',
        absolute_tolerance: float = 1e-8,
    ) -> bool:
        return bool(
            torch.isclose(
                t1.rotation_and_translation,
                t2.rotation_and_translation,
                atol=absolute_tolerance,
            ).all()
        )

    # Cannot write `IDENTITY: ClassVar[Self] = Transformation.from_parts`.
    # Class properties are also deprecated.
    # Python... :v
    @classmethod
    def identity(cls) -> Self:
        return cls.from_parts(
            torch.eye(3, 3, dtype=torch.float32),
            torch.zeros(3, dtype=torch.float32),
        )

    @classmethod
    def from_parts(
        cls,
        rotation: Float[torch.Tensor, '3 3'],
        translation: Float[torch.Tensor, '3'],
    ) -> Self:
        rotation_and_translation = torch.zeros(
            (4, 4),
            dtype=rotation.dtype,
            device=rotation.device,
        )
        rotation_and_translation[:3, :3] = rotation
        rotation_and_translation[:3, 3] = translation
        rotation_and_translation[3, 3] = 1.0

        return cls(rotation_and_translation)

    @classmethod
    def active(
        cls,
        intrinsic_euler_angles: Float[torch.Tensor, '3'] | None = None,
        translation: Float[torch.Tensor, '3'] | None = None,
    ) -> Self:
        rotation = (
            (
                rotation_around('x', intrinsic_euler_angles[0])
                @ rotation_around('y', intrinsic_euler_angles[1])
                @ rotation_around('z', intrinsic_euler_angles[2])
            )
            if intrinsic_euler_angles is not None
            else torch.eye(3)
        )

        translation = translation if translation is not None else torch.zeros(3)

        return cls.from_parts(rotation, translation)

    def clone(self) -> 'Transformation':
        return Transformation(self.rotation_and_translation.clone())

    def cpu(self) -> 'Transformation':
        return Transformation(self.rotation_and_translation.cpu())

    def to(
        self,
        device: torch.device | None = None,
        dtype: torch.dtype | None = None,
    ) -> 'Transformation':
        return Transformation(self.rotation_and_translation.to(device, dtype, copy=True))
