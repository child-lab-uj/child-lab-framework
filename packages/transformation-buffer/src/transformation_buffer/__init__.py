from .buffer import Buffer
from .transformation import Transformation

__all__ = ['Buffer', 'Transformation']
