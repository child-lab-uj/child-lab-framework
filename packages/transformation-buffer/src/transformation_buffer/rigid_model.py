from math import pi
from typing import Hashable, Protocol, runtime_checkable

import torch
from attrs import frozen

from transformation_buffer.transformation import Transformation


@runtime_checkable
class RigidModel[T: Hashable](Protocol):
    def transformations(self) -> dict[tuple[T, T], Transformation]: ...


@frozen
class Cube[T: Hashable]:
    width: float
    walls: tuple[T, T, T, T, T, T]

    # Cube Walls:
    #    4
    #  5 1 6
    #    2
    #    3

    def transformations(self) -> dict[tuple[T, T], Transformation]:
        first, second, third, forth, fifth, sixth = self.walls

        half_pi = pi / 2.0
        half_width = self.width / 2.0

        dtype = torch.float64

        down = Transformation.active(
            intrinsic_euler_angles=torch.tensor((-half_pi, 0.0, 0.0)),
            translation=torch.tensor((0.0, half_width, -half_width)),
        ).to(dtype=dtype)

        left = Transformation.active(
            intrinsic_euler_angles=torch.tensor((0.0, half_pi, 0.0)),
            translation=torch.tensor((half_width, 0.0, -half_width)),
        ).to(dtype=dtype)

        right = Transformation.active(
            intrinsic_euler_angles=torch.tensor((0.0, -half_pi, 0.0)),
            translation=torch.tensor((-half_width, 0.0, -half_width)),
        ).to(dtype=dtype)

        return {
            (first, second): down.clone(),
            (second, third): down.clone(),
            (third, forth): down,
            (first, fifth): left,
            (first, sixth): right,
        }
