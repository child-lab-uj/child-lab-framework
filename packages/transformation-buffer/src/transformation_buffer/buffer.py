from collections.abc import Callable, Hashable, Iterable
from copy import deepcopy
from functools import reduce
from typing import Any, Self

import serde
from more_itertools import pairwise
from plum import dispatch, overload
from rustworkx import (
    PyDiGraph,
    digraph_dijkstra_search,
    digraph_dijkstra_shortest_paths,
    is_connected,
)
from rustworkx.visit import DijkstraVisitor

from transformation_buffer.rigid_model import RigidModel
from transformation_buffer.serialization import graph, tensor
from transformation_buffer.transformation import Transformation

# Initialize custom global class serializers & deserializers.
graph.init()
tensor.init()

# TODO: Implement path compression and verification.


@serde.serde
class Buffer[T: Hashable]:
    __frames_of_reference: set[T]
    __frame_names_to_node_indices: dict[T, int]
    __connections: PyDiGraph[T, Transformation] = serde.field(
        serializer=graph.serialize_graph,
        deserializer=graph.deserialize_graph,
    )

    @overload
    def __init__(self, frames_of_reference: Iterable[T] | None = None) -> None:  # noqa: F811
        nodes = (
            set() if frames_of_reference is None else deepcopy(set(frames_of_reference))
        )
        n_nodes = len(nodes)

        connections = PyDiGraph[T, Transformation](
            multigraph=False,  # Think about it later
            node_count_hint=n_nodes,
            edge_count_hint=n_nodes,
        )

        node_indices = connections.add_nodes_from(nodes)

        self.__frames_of_reference = nodes
        self.__frame_names_to_node_indices = dict(zip(nodes, node_indices))
        self.__connections = connections

    @overload
    def __init__(  # noqa: F811
        self,
        frames_of_reference: set[T],
        frame_names_to_node_indices: dict[T, int],
        connections: PyDiGraph[T, Transformation],
    ) -> None:
        self.__frames_of_reference = frames_of_reference
        self.__frame_names_to_node_indices = frame_names_to_node_indices
        self.__connections = connections

    @dispatch
    def __init__(self, *args: Any, **kwargs: Any) -> None:  # noqa: F811
        pass

    @property
    def connected(self) -> bool:
        return is_connected(self.__connections.to_undirected(multigraph=False))

    @property
    def frames_of_reference(self) -> frozenset[T]:
        # Construction of frozenset from set of strings is *somehow* optimized.
        # Perhaps there's no need for caching this property.
        return frozenset(self.__frames_of_reference)

    def add_object(self, model: RigidModel[T]) -> Self:
        for from_to, transformation in model.transformations().items():
            self[from_to] = transformation

        return self

    def add_frame_of_reference(self, frame: T) -> Self:
        self.__add_frame_of_reference(frame)
        return self

    def add_transformation(
        self,
        from_frame: T,
        to_frame: T,
        transformation: Transformation,
    ) -> Self:
        if self[from_frame, to_frame] is not None:
            return self

        self[from_frame, to_frame] = transformation

        return self

    def frames_visible_from(self, frame: T) -> list[T]:
        if (frame_index := self.__frame_index(frame)) is None:
            return []

        class Visitor(DijkstraVisitor):
            visible_nodes: list[int] = []

            def discover_vertex(self, v: int, score: float) -> None:
                self.visible_nodes.append(v)

        visitor = Visitor()
        digraph_dijkstra_search(self.__connections, [frame_index], visitor=visitor)

        visible_node_indices = visitor.visible_nodes
        visible_node_indices.remove(frame_index)

        return [self.__connections.get_node_data(node) for node in visible_node_indices]

    def __add_frame_of_reference(self, frame: T) -> int:
        if (index := self.__frame_index(frame)) is not None:
            return index

        self.__frames_of_reference.add(frame)
        index = self.__connections.add_node(frame)
        self.__frame_names_to_node_indices[frame] = index

        return index

    def __frame_index(self, frame: T) -> int | None:
        return self.__frame_names_to_node_indices.get(frame, None)

    def __setitem__(self, from_to: tuple[T, T], transformation: Transformation) -> None:
        from_id, to_id = from_to

        from_index = self.__frame_index(from_id) or self.__add_frame_of_reference(from_id)
        to_index = self.__frame_index(to_id) or self.__add_frame_of_reference(to_id)

        self.__connections.add_edge(from_index, to_index, transformation)
        self.__connections.add_edge(to_index, from_index, transformation.inverse())

    def __getitem__(self, from_to: tuple[T, T]) -> Transformation | None:
        from_id, to_id = from_to

        from_index = self.__frame_index(from_id)
        to_index = self.__frame_index(to_id)

        if from_index is None or to_index is None:
            return None

        if from_index == to_index:
            return Transformation.identity().clone()

        connections = self.__connections

        if self.__connections.has_edge(from_index, to_index):
            return self.__connections.get_edge_data(from_index, to_index)

        path_mapping = digraph_dijkstra_shortest_paths(
            connections,
            from_index,
            to_index,
        )

        if to_index not in path_mapping:
            return None

        shortest_path = path_mapping[to_index]

        return map_reduce(  # type: ignore[no-any-return]  # MyPy complains about the possible `Any`.
            lambda nodes: connections.get_edge_data(*nodes),
            lambda t1, t2: t1 @ t2,
            pairwise(shortest_path),
        )

    def __contains__(self, from_to: tuple[T, T]) -> bool:
        from_id, to_id = from_to

        from_index = self.__frame_index(from_id)
        to_index = self.__frame_index(to_id)

        if from_index is None or to_index is None:
            return False

        return self.__connections.has_edge(from_index, to_index)


def map_reduce[U, V](
    map_function: Callable[[U], V],
    reduce_function: Callable[[V, V], V],
    iterable: Iterable[U],
) -> V:
    return reduce(reduce_function, map(map_function, iterable))
