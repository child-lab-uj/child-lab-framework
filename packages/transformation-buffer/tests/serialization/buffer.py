from pathlib import Path

import pytest
from serde.json import from_json, to_json
from syrupy.assertion import SnapshotAssertion
from transformation_buffer import Buffer, Transformation


@pytest.fixture
def empty_buffer() -> Buffer[str]:
    return Buffer[str]()


@pytest.fixture
def buffer() -> Buffer[str]:
    return (
        Buffer[str]()
        .add_transformation('a', 'b', Transformation.identity())
        .add_transformation('a', 'c', Transformation.identity())
        .add_transformation('b', 'd', Transformation.identity())
        .add_transformation('d', 'e', Transformation.identity())
        .add_transformation('c', 'e', Transformation.identity())
        .add_transformation('d', 'f', Transformation.identity())
        .add_transformation('g', 'h', Transformation.identity())
    )


def test_serialize_empty_buffer_json(
    empty_buffer: Buffer[str],
    snapshot: SnapshotAssertion,
) -> None:
    serialized = to_json(empty_buffer)
    assert serialized == snapshot


def test_serialize_buffer_json(
    buffer: Buffer[str],
    snapshot: SnapshotAssertion,
) -> None:
    serialized = to_json(buffer)
    assert serialized == snapshot


def test_deserialize_buffer_json() -> None:
    data_location = Path(__file__).parent / 'data' / 'serialized_buffer.json'
    data = data_location.read_text()

    deserialized = from_json(Buffer[str], data)
    assert deserialized['a', 'b'] == Transformation.identity()
