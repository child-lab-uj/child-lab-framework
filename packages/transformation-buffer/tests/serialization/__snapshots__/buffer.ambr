# serializer version: 1
# name: test_serialize_buffer_json
  '{"_Buffer__frames_of_reference":["c","h","g","a","f","e","d","b"],"_Buffer__frame_names_to_node_indices":{"a":0,"b":1,"c":2,"d":3,"e":4,"f":5,"g":6,"h":7},"_Buffer__connections":{"check_cycle":false,"multigraph":false,"nodes":[[0,"a"],[1,"b"],[2,"c"],[3,"d"],[4,"e"],[5,"f"],[6,"g"],[7,"h"]],"edges":[[[0,1],{"rotation_and_translation":{"data":[[1.0,0.0,0.0,0.0],[0.0,1.0,0.0,0.0],[0.0,0.0,1.0,0.0],[0.0,0.0,0.0,1.0]],"dtype":"torch.float32"}}],[[1,0],{"rotation_and_translation":{"data":[[1.0,0.0,0.0,0.0],[0.0,1.0,0.0,0.0],[0.0,0.0,1.0,0.0],[0.0,0.0,0.0,1.0]],"dtype":"torch.float32"}}],[[0,2],{"rotation_and_translation":{"data":[[1.0,0.0,0.0,0.0],[0.0,1.0,0.0,0.0],[0.0,0.0,1.0,0.0],[0.0,0.0,0.0,1.0]],"dtype":"torch.float32"}}],[[2,0],{"rotation_and_translation":{"data":[[1.0,0.0,0.0,0.0],[0.0,1.0,0.0,0.0],[0.0,0.0,1.0,0.0],[0.0,0.0,0.0,1.0]],"dtype":"torch.float32"}}],[[1,3],{"rotation_and_translation":{"data":[[1.0,0.0,0.0,0.0],[0.0,1.0,0.0,0.0],[0.0,0.0,1.0,0.0],[0.0,0.0,0.0,1.0]],"dtype":"torch.float32"}}],[[3,1],{"rotation_and_translation":{"data":[[1.0,0.0,0.0,0.0],[0.0,1.0,0.0,0.0],[0.0,0.0,1.0,0.0],[0.0,0.0,0.0,1.0]],"dtype":"torch.float32"}}],[[3,4],{"rotation_and_translation":{"data":[[1.0,0.0,0.0,0.0],[0.0,1.0,0.0,0.0],[0.0,0.0,1.0,0.0],[0.0,0.0,0.0,1.0]],"dtype":"torch.float32"}}],[[4,3],{"rotation_and_translation":{"data":[[1.0,0.0,0.0,0.0],[0.0,1.0,0.0,0.0],[0.0,0.0,1.0,0.0],[0.0,0.0,0.0,1.0]],"dtype":"torch.float32"}}],[[3,5],{"rotation_and_translation":{"data":[[1.0,0.0,0.0,0.0],[0.0,1.0,0.0,0.0],[0.0,0.0,1.0,0.0],[0.0,0.0,0.0,1.0]],"dtype":"torch.float32"}}],[[5,3],{"rotation_and_translation":{"data":[[1.0,0.0,0.0,0.0],[0.0,1.0,0.0,0.0],[0.0,0.0,1.0,0.0],[0.0,0.0,0.0,1.0]],"dtype":"torch.float32"}}],[[6,7],{"rotation_and_translation":{"data":[[1.0,0.0,0.0,0.0],[0.0,1.0,0.0,0.0],[0.0,0.0,1.0,0.0],[0.0,0.0,0.0,1.0]],"dtype":"torch.float32"}}],[[7,6],{"rotation_and_translation":{"data":[[1.0,0.0,0.0,0.0],[0.0,1.0,0.0,0.0],[0.0,0.0,1.0,0.0],[0.0,0.0,0.0,1.0]],"dtype":"torch.float32"}}]]}}'
# ---
# name: test_serialize_empty_buffer_json
  '{"_Buffer__frames_of_reference":[],"_Buffer__frame_names_to_node_indices":{},"_Buffer__connections":{"check_cycle":false,"multigraph":false,"nodes":[],"edges":[]}}'
# ---
